### ISSUE TITLE
About hyperparameter optimization

### ISSUE THREAD

**Issue by moskomule** [AlexImmer/marglik#1]
Hi, I really enjoyed your impressive paper.

I read `marglik.py` and tried to replicate the optimization of the temperature parameter written in the paper as eq (7). But I found the returned value of `lap.log_marginal_likelihood` is not differentiable w.r.t. parameters other than the prior precision, that is, the temperature parameter cannot be optimized. It would be appreciated if you instruct me on how to do this.

**Comment by aleximmer** [comment]
Thanks @moskomule! I will get back to you in 1-2 weeks, I am currently out of office.

**Comment by aleximmer** [comment]
Unfortunately, the current code does not allow to do the exact same hyperparameter optimization as done in the paper. For the experiment in the paper, I implemented a custom likelihood that would scale the logits (neural network function before softmax) by a hyperparameter, let's call it `alpha`. `alpha` then impacts the likelihood directly and this is not possible using the second-order backends of the Laplace library.

However, here is a way to obtain a similar result: we will add the same temperature parameter `alpha` to the neural network itself. That means, we basically add one final parameter to the neural network that is just a scale `alpha` and has a mean of `1` (!) and we optimize it like every other parameter with another prior precision hyperparameter on this `alpha` that determines its prior precision. To do this, only the `AsdlGGN` or `AsdlEF` backends work. Then, simply set up a neural network as follows:
```python
from asdfghjkl.operations import Scale 
model = nn.Sequential(nn.Linear(2, 50), nn.Tanh(), nn.Linear(50, 2), Scale())  # Scale() is `alpha`
prior_mean = torch.zeros_like(parameters_to_vector(model.parameters()))
prior_mean[-1] = 1.0  # `alpha` prior should be mean 1 which means no temperature scaling
```

I hope this helps. Otherwise let me know if something is unclear or you have more questions.

**Comment by moskomule** [comment]
Thank you! I hope the implementation of the paper will be included in this repository in the future.
